lib/gc/Dialect/Linalgx/LinalgxOps.cpp

```diff
@@ -82,6 +82,517 @@ void SigmoidOp::getEffects(
+                     .getValue());
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  cached = Builder(context).getAffineMapArrayAttr(maps);
+  getOperation()->setAttr(memoizeAttr, cached);
+  return cached;
+}
+
+void Mm2DVnniOp::regionBuilder(ImplicitLocOpBuilder &b, Block &block,
+                               ArrayRef<NamedAttribute> attrs) {
+  assert(3 > 0 && block.getNumArguments() == 3 &&
+         "Mm2DVnniOp regionBuilder expects 3 (>=0) args");
+  RegionBuilderHelper helper(b, block);
+  SmallVector<Value> yields;
+
+  Value value1 =
+      helper.buildTypeFn(TypeFn::cast_signed, block.getArgument(2).getType(),
+                         block.getArgument(0));
+  Value value2 =
+      helper.buildTypeFn(TypeFn::cast_signed, block.getArgument(2).getType(),
+                         block.getArgument(1));
+  Value value3 = helper.buildBinaryFn(BinaryFn::mul, value1, value2);
+  Value value4 =
+      helper.buildBinaryFn(BinaryFn::add, block.getArgument(2), value3);
+  yields.push_back(value4);
+  helper.yieldOutputs(yields);
+}
+
+ParseResult Mm2DVnniOp::parse(OpAsmParser &parser, OperationState &result) {
+  return ::parseNamedStructuredOp(parser, result,
+                                  Mm2DVnniOp::getNumRegionArgs(),
+                                  Mm2DVnniOp::getRegionBuilder());
+}
+
+void Mm2DVnniOp::print(OpAsmPrinter &p) {
+  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs());
+}
+
+LogicalResult Mm2DVnniOp::fold(FoldAdaptor, SmallVectorImpl<OpFoldResult> &) {
+  return memref::foldMemRefCast(*this);
+}
+
+void Mm2DVnniOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  if (hasPureTensorSemantics())
+    return;
+  getGenericEffectsImpl(effects, getOperation()->getResults(), getDpsInputs(),
+                        getDpsInits());
+}
+
+LogicalResult Mm2DVnniOp::verify() {
+  // A[M, K]
+  // B[N0, K0, K1, N1, K2]
+  // C[M, N]
+  auto shapeA = ShapeAdaptor(getInputs()[0].getType());
+  auto shapeB = ShapeAdaptor(getInputs()[1].getType());
+  auto shapeC = ShapeAdaptor(getOutputs()[0].getType());
+  // check rank
+  auto hasRank = shapeA.hasRank() && shapeB.hasRank() && shapeC.hasRank();
+  if (!hasRank)
+    return emitOpError() << "input/output must have rank.";
+  auto checkRank = (shapeA.getRank() == 2) && (shapeB.getRank() == 5) &&
+                   (shapeC.getRank() == 2);
+  if (!checkRank)
+    return emitOpError() << "not supported input/output shape.";
+  // match M, N, K dims
+  bool matchM = shapeA.getDimSize(0) == shapeC.getDimSize(0);
+  bool matchN =
+      (shapeB.getDimSize(0) * shapeB.getDimSize(3)) == shapeC.getDimSize(1);
+  bool matchK =
+      shapeA.getDimSize(1) ==
+      (shapeB.getDimSize(1) * shapeB.getDimSize(2) * shapeB.getDimSize(4));
+  bool matchVnni = (shapeB.getDimSize(4) == 2) || (shapeB.getDimSize(4) == 4);
```

Comment:
Shall we also check vnni dim value based on dtype? (e.g. we restrain vnni to be 2 under bf16; and 4 under u8/s8).

Reply:
Added.